pkg/resources/management.cattle.io/v3/roletemplate/validator.go

```diff
@@ -73,65 +81,125 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
-				Reason:  metav1.StatusReasonBadRequest,
-				Code:    http.StatusBadRequest,
-			},
-			Allowed: false,
-		}, nil
-	}
-
-	rules, err := a.roleTemplateResolver.RulesFromTemplate(roleTemplate)
+		return admission.ResponseBadRequest(fmt.Sprintf("Circular Reference: RoleTemplate %s already inherits RoleTemplate %s", circularTemplate.Name, newRT.Name)), nil
+	}
+
+	rules, err := a.roleTemplateResolver.RulesFromTemplate(newRT)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get all rules for '%s': %w", newRT.Name, err)
 	}
 
-	// ensure all PolicyRules have at least one verb, otherwise RBAC controllers may encounter issues when creating Roles and ClusterRoles
-	for i := range rules {
-		if len(rules[i].Verbs) == 0 {
-			return &admissionv1.AdmissionResponse{
-				Result: &metav1.Status{
-					Status:  "Failure",
-					Message: "RoleTemplate.Rules: PolicyRules must have at least one verb",
-					Reason:  metav1.StatusReasonBadRequest,
-					Code:    http.StatusBadRequest,
-				},
-				Allowed: false,
-			}, nil
-		}
+	// verify inherited rules have verbs
+	if err := common.CheckForVerbs(rules); err != nil {
+		return admission.ResponseBadRequest(err.Error()), nil
 	}
 
-	allowed, err := auth.EscalationAuthorized(request, roleTemplateGVR, a.sar, "")
+	allowed, err := auth.EscalationAuthorized(request, gvr, a.sar, "")
 	if err != nil {
 		logrus.Warnf("Failed to check for the 'escalate' verb on RoleTemplates: %v", err)
+	} else if allowed {
+		return admission.ResponseAllowed(), nil
 	}
 
-	if allowed {
-		return &admissionv1.AdmissionResponse{Allowed: true}, nil
+	err = auth.ConfirmNoEscalation(request, rules, "", a.resolver)
+	if err != nil {
+		return admission.ResponseFailedEscalation(err.Error()), nil
 	}
-	response := &admissionv1.AdmissionResponse{}
-	auth.SetEscalationResponse(response, auth.ConfirmNoEscalation(request, rules, "", a.resolver))
-	return response, nil
+
+	return admission.ResponseAllowed(), nil
+}
+
+// validateUpdateFields checks if the fields being changed are valid update fields.
+func (a *admitter) validateUpdateFields(oldRole, newRole *v3.RoleTemplate, fldPath *field.Path) *field.Error {
+	if err := validateContextValue(newRole, fldPath); err != nil {
+		return err
+	}
+
+	// if this is not a built in role no further validation is needed
+	if !oldRole.Builtin {
+		return nil
+	}
+
+	// allow changes to meta data and defaults
+	oldRole.ClusterCreatorDefault = newRole.ClusterCreatorDefault
+	oldRole.ProjectCreatorDefault = newRole.ProjectCreatorDefault
+
+	// we do not want to block K8s controllers from adding metadata or typemeta to the object.
+	oldRole.ObjectMeta = newRole.ObjectMeta
+	oldRole.TypeMeta = newRole.TypeMeta
+
+	if !equality.Semantic.DeepEqual(oldRole, newRole) {
+		return field.Forbidden(fldPath, "updates to builtIn RoleTemplates for fields other than CreatorDefault are forbidden")
```

Comment:
This looks like a behavior change from https://github.com/rancher/rancher/blob/64c42a24379600b3061aa014ec1d25147a77699f/pkg/api/norman/customization/roletemplate/roletemplate.go#L42-L51 . In the norman version, changed fields seem to be silently dropped. Is the change intentional?